doc/enter.1.in
.TH ENTER "1" "December 2006" "enter @VERSION@" "User Commands"
.SH NAME
Enter @VERSION@ \- Lightweight graphical login manager
.SH SYNOPSIS
.TP 5
\fBenter \fI[options]\fR

.SH DESCRIPTION
.PP
Enter is a lightweight login manager for X.
.TP
\fB\-c\fR
specify an alternative config file
.TP
\fB\-a\fR
specify an alternative auth file
.TP
\fB\-l\fR
specify an alternative login file
.TP
\fB\-d\fR
specify which display to connect to
.TP
\fB\-t\fR
use specified theme
.TP
\fB\-n\fR
run in foreground
.TP
\fB\-v\fR
print version information
.TP
\fB\-h\fR
print help information

.SH FILES
@sysconfdir@/enter.conf
.br
@datadir@/@PACKAGE@/themes

.SH "REPORTING BUGS"
Report bugs to <rob_dobbs@users.sourceforge.net>.

.SH AUTHOR
Written by Robert Knutsson.

.SH "SEE ALSO"
\fIhttp://enter.sourceforge.net\fR,
\fIenter.conf(5)\fR


ChangeLog
2007-05-16  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/greeter.c:
	code from old gui.c.
	* src/gui.c:
	stripped all high-level code.
	* src/list.c:
	implemented a linked list.
	* src/conf.c:
	conf_t is now a linked list of type list_t.

2007-05-08  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui.c:
	removed Background pixmap from struct, since we can free
	it as soon as it has been passed to SetBackground.

	* src/enter.c:
	arguments passed to the program now overrides options in
	config file.

2007-05-07  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/conf.c:
	added const strings and strdup to address a bug
	caused by memory corruption.

2007-05-05  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/conf.c:
	removed const strings.

	* src/enter.c:
	added errno.h.

2007-05-04  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/conf.[ch]:
	Removed map_t typedef.

2007-04-11  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/:
	All files where spell-checked.

2007-03-31  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/:
	All log priorities where adjusted.

2007-03-30  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/login.c:
	shadow is now optional, will only be compiled
	if the header shadow.h is found.

2007-03-26  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/login.c:
	converts relative to absolute pathnames.

2007-03-26  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/login.c:
	enter now has seperate auth files for
	server and client.
	* src/enter.c:
	enter.conf can specify which authentication
	protocol to use.

2007-03-13  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/auth.c:
	added auth.[ch] to SVN, sorry :S
	added X authentication code here.

2007-03-09  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/auth.[ch]:
	moved to login.[ch].

2007-03-02  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/server.c:
	added MIT cookie generator code.

2007-03-01  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui_input.c:
	added code to show the cursor.
	* TODO:
	enter dont seem to loose focus anymore when run
	at system start.

2007-03-01  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui_input.c:
	cursor code now works.
	added code to move cursor.

2007-03-01  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui_input.c:
	worked on cursor code.
	* src/gui.c:
	now uses gui_input.c's new code.
	* data/themes/default/theme:
	tweaked the values.

2007-02-27  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui_input.c:
	added basic code for cursor functions.
	* data/themes/default:
	changed default theme.
	* src/gui.c:
	changed theme syntax.
	* src/auth.c:
	auth_login now takes the login_file and auth_file as arguments
	* src/enter.c:
	login_file and auth_file now gets default values.
	* doc/:
	updated the documentation.

2007-01-10  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/Makefile.am:
	changed install path for binaries to bin instead of sbin.

2006-12-23  Robert Knutsson <rob_dobbs@users.sf.net>
	* doc:
	finished writing the manuals.
	* src/auth.c:
	included crypt.h, and got rid of stupid warning.
	* README, AUTHORS:
	added content.
	* src/gui.c:
	added new member `drawable' which points to the drawable
	surface, doublebuffering or not.

2006-12-21  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/auth:
	removed dependency on conf.h
	* src/enter.c:
	removed the `display' entry from theme.
	* doc:
	added man pages, enter.1 and enter.conf.5.
	* src/display, src/enter:
	now kills clients after a user session.

2006-12-20  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui.c:
	added doublebuffering support through Xdbe.
	* src/auth.c:
	fixed a double-free bug.
	* src/enter.c:
	enter.conf now only contains the theme name.
	* src/gui_widget.c:
	fixed problem with incomplete type.

2006-12-11  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/auth.c:
	added a check mechanism to ensure a user was
	successfuly authenticated before a login.
	* src/auth.h, src/display.h, src/conf.h:
	commented the function prototypes.
	* autoconf.ac:
	replaced version with "svn", this should only be set
	when tagging a release.

2006-12-08  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/greeter.c:
	removed greeter, moved the code to enter.c instead.
	* src/display.c:
	added function `display_kill_clients'.

2006-12-06  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/greeter.c:
	changed greeter from a singleton to a regular class.
	the connection to the X display is now open during a
	user session.
	* src/server.c:
	added code to shutdown the server.
	* src/enter.c:
	now kills the server when the greeter fails to load.

2006-12-04  Robert Knutsson <rob_dobbs@users.sf.net>
	* configure.ac:
	fixed a bug with x_includes.
	general cleanup.

2006-12-01  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/gui_widgets.c:
	splitted into gui_input.c and gui_label.c
	* src/enter.c:
	dont need root to display help no more.
	* src/auth.c:
	changed to singleton.

2006-11-30  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/Makefile.am:
	only compile one binary instead of two seperate.
	* src/greeter_*:
	removed the greeter_ part of the filename.
	* src/gui.c:
	displays error msg at unsuccessful login.
	fixed bug when `enter.visible' isn't `both'.
	added include "X11/keysymdef.h".

2006-11-27  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/greeter_image.c:
	renamed file to greeter_gui_image.c,
	and renamed image_t to gui_image_t.

2006-11-27  Robert Knutsson <rob_dobbs@users.sf.net>
	* src/greeter_image.c:
	renamed image_free to image_delete.
	* src/greeter_gui.c:
	handle MappingNotify XEvent.
	* src/greeter_gui_widgets.c:
	removed duplicate code in gui_input_t,
	which now uses gui_label_t.

2006-11-26  Robert Knutsson <rob_dobbs@users.sf.net>
	* data/themes/default/input.png:
	added transparency.

2006-11-25  Robert Knutsson <rob_dobbs@users.sf.net>
	* greeter_gui.c:
	deletes window in gui_delete
	* greeter_display.c:
	reads DISPLAY from config instead of environment.
	* greeter.c, greeter_auth.c:

src/enter.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>
#include <errno.h>

#include <X11/Xauth.h>

#include "enter.h"
#include "server.h"
#include "gui.h"
#include "log.h"
#include "conf.h"
#include "utils.h"
#include "login.h"
#include "auth.h"

#define PIDFILE "/var/run/" PACKAGE ".pid"
#define PIDBUF 20
#define THEME_LEN 512

static void parse_args(int argc, char **argv, conf_t *conf)
{
	int i;

	for (i=1;i<argc;i++) {
		if ((strcmp(argv[i],"-c") == 0) && (i+1 < argc)) {
			conf_set(conf,"config_file",argv[++i]);

		} else if ((strcmp(argv[i],"-a") == 0) && (i+1 < argc)) {
			conf_set(conf,"auth_file",argv[++i]);

		} else if ((strcmp(argv[i],"-l") == 0) && (i+1 < argc)) {
			conf_set(conf,"login_file",argv[++i]);
			
		} else if ((strcmp(argv[i],"-d") == 0) && (i+1 < argc)) {
			conf_set(conf,"display",argv[++i]);
		
		} else if ((strcmp(argv[i],"-t") == 0) && (i+1 < argc)) {
			conf_set(conf,"theme",argv[++i]);

		} else if (strcmp(argv[i],"-n") == 0) {
			conf_set(conf,"daemon","false");

		} else if (strcmp(argv[i],"-v") == 0) {
			printf("%s version %s\n", PACKAGE, VERSION);
			exit(EXIT_SUCCESS);

		} else if (strcmp(argv[i],"-h") == 0) {
			printf(
			"Usage: %s: [OPTIONS]\n\n"
			"  -c CONFIG   use config file CONFIG\n"
			"  -a AUTH     write server authentication to AUTH\n"
			"  -l LOGIN    run LOGIN as the user session\n"
			"  -d DISPLAY  connect to display DISPLAY\n"
			"  -t THEME    override config theme's setting with THEME\n"
			"  -n          dont run as a daemon\n"
			"  -v          print version information\n"
			"  -h          print this help info\n"
			"\n"
			"Report bugs to <%s>.\n",
			argv[0], PACKAGE_BUGREPORT);
			exit(EXIT_SUCCESS);
		} else {
			printf("unknown argument, try -h\n");
			exit(EXIT_SUCCESS);
		}
	}
}

static void default_settings(conf_t *conf)
{
	conf_set(conf, "auth_file", "/tmp/enter.xauth");
	conf_set(conf, "login_file", ".xinitrc");
	conf_set(conf, "daemon", "true");
	conf_set(conf, "authentication", "magic-cookie");
	conf_set(conf, "display", ":0");
	conf_set(conf, "theme", "default");
}

static int daemonize()
{
	pid_t pid, sid;
	pid = fork();
	if (pid < 0) {
		log_print(LOG_ERR, "Could not fork to background.");
		return FALSE;
	} else if (pid > 0) {
		/* Kill the parent thread.  */
		exit(EXIT_SUCCESS);
	}
	
	sid = setsid();
	
	if (sid < 0) {
		log_print(LOG_ERR, "Could not set session id.");
		return FALSE;
	}
	
	if (chdir("/") < 0) {
		log_print(LOG_ERR, "Could not change working directory.");
		return FALSE;
	}

	/* Do not use `close' as this causes broken pipe's with the children.  */
	freopen("/dev/null", "r", stdin);
	freopen("/dev/null", "w", stdout);
	freopen("/dev/null", "w", stderr);

	log_daemon(TRUE);

	return TRUE;
}

static auth_t* setup_authentication(conf_t *conf)
{
	auth_t *auth = NULL;
	char hostname[128];
	char *display = rindex(conf_get(conf, "display"), ':');

	/* If the colon wasn't found, set the display to a default value,
	 * otherwise increase display to point beyond the colon.  */
	if (display) {
		display++;
	} else {
		display = "0";
	}

	memset(hostname, '\0', 128);
	gethostname(hostname, 127);

	char *authentication = conf_get(conf, "authentication");

	if (!strcmp(authentication, "magic-cookie")) {
		auth = auth_new(AUTH_MIT_MAGIC_COOKIE, hostname, display);
	} else if (!strcmp(authentication, "xdm")) {
		log_print(LOG_ERR, "XDM is currently not implemented.");
		return NULL;
	} else {
		log_print(LOG_ERR, "'%s' is a unknown authentication protocol.", 
				authentication);
		return NULL;
	}
		
	if (!auth) {
		return NULL;
	}

	if (unlink(conf_get(conf, "auth_file")) == -1) {
		log_print(LOG_WARNING, "Could not remove auth file: %s.",
				strerror(errno));
	}

	if (!auth_write(auth, conf_get(conf, "auth_file"))) {
		log_print(LOG_ERR, "Could not write to auth file.");
		return NULL;
	}
	
	setenv("XAUTHORITY", conf_get(conf, "auth_file"), 1);

	return auth;
}

static void write_pidfile(pid_t pid)
{
	char buf[PIDBUF];
	int length;
	FILE *fp = fopen(PIDFILE,"w");

	if (!fp) {
		log_print(LOG_ERR, "Could not write pidfile: %s.",PIDFILE);
		return;
	}

	memset(buf,'\0',PIDBUF);
	length = snprintf(buf,PIDBUF-1,"%d",pid);
	
	fwrite(buf,sizeof(char),length,fp);
	fclose(fp);

	return;
}

static void remove_pidfile()
{
	remove(PIDFILE);
}

static void shutdown()
{
	remove_pidfile();
	/* TODO: remove auth file.  */
}

int main(int argc, char **argv)
{
	conf_t *conf = conf_new();
	conf_t *theme = conf_new();
	conf_t *cmd = conf_new();
	auth_t *auth = NULL;
	
	/* Assign default settings to conf
	 * and parse command line arguments.  */
	default_settings(conf);

	conf_set(cmd,"config_file", CONFDIR "/enter.conf");
	parse_args(argc, argv, cmd);

	/* Check if we have enough privileges.  */
	if (getuid() != 0) {
		log_print(LOG_EMERG, "Not enough privileges to run.");
		exit(EXIT_FAILURE);
	}

	/* If the auth file is going to be shared,
	 * umask needs to be set so everyone has reading
	 * permissions on the files written. Otherwise umask
	 * should be set so only the creator can read and write.  */
	umask(077);
	
	openlog(PACKAGE, LOG_NOWAIT, LOG_DAEMON);
	
	/* Parse config file.  */
	if (conf_parse(conf, conf_get(cmd,"config_file")) == FALSE) {
		log_print(LOG_EMERG, "Could not read config file: \"%s\"",
					conf_get(cmd, "config_file"));
		closelog();
		exit(EXIT_FAILURE);
	}

	conf_merge(conf, cmd);
	conf_delete(cmd);

	/* Parse theme file.  */
	char theme_file[THEME_LEN];
	snprintf(theme_file, THEME_LEN-1, "%s/%s/theme", THEMEDIR,
			conf_get(conf, "theme"));

	if (conf_parse(theme, theme_file) == FALSE) {
		log_print(LOG_EMERG,
			"Could not parse theme \"%s\"",
			theme_file);
		closelog();
		exit(EXIT_FAILURE);
	}

	/* Add theme name to theme, so the full path of data files
	 * can be read by only supplying theme.  */
	conf_set(theme, "theme",
			conf_get(conf, "theme"));

	/* Fork to background if "daemon" mode is enabled in the config. */
	if (!strcmp(conf_get(conf,"daemon"),"true")) {
		if (daemonize() == FALSE) {
			log_print(LOG_EMERG, "Could not daemonize enter.");
			closelog();
			exit(EXIT_FAILURE);
		}
	}
	
	write_pidfile(getpid());
	atexit(shutdown);

	if (strcmp(conf_get(conf, "authentication"), "none") != 0) {
		auth = setup_authentication(conf);
		if (!auth) {
			log_print(LOG_ERR, "Could not setup authentication system, disabling authentication.");
			conf_set(conf, "authentication", "none");
		}
	}

	log_print(LOG_DEBUG, "Starting X server.");
	pid_t server_pid = server_start(conf);
	if (server_pid == FALSE) {
		log_print(LOG_EMERG, "Could not start X server");
		closelog();
		exit(EXIT_FAILURE);
	}

	/* Connect to the X display.  */
	display_t *display = display_new(conf);
	if (!display) {
		log_print(LOG_EMERG, "Could not connect to X display.");
		server_stop();
		closelog();
		exit(EXIT_FAILURE);
	}
	
	/* Create a GUI window.  */
	gui_t *gui = gui_new(display);
	if (!gui) {
		log_print(LOG_EMERG, "Could not open GUI.");
		server_stop();
		closelog();
		exit(EXIT_FAILURE);
	}

	while (1) {
#if 0
		gui_show(gui);
		int action = gui_run(gui);
		gui_hide(gui);

		switch (action) {
		case LOGIN:
			log_print(LOG_DEBUG, "Logging in user...");

			if (login_start_session(conf_get(conf, "display"),
						auth, ".Xauthority",
						conf_get(conf, "login_file")) == FALSE) {
				log_print(LOG_EMERG,
						"Could not open user session.");
				
				server_stop();
				closelog();
				exit(EXIT_FAILURE);
			}
			break;
		}
		display_kill_clients(display, gui->win);
#endif
	}


	log_print(LOG_DEBUG, "Shutting down.");
	
	gui_delete(gui);
	conf_delete(conf);
	conf_delete(theme);
	
	server_stop();
	closelog();
	
	exit(EXIT_SUCCESS);
}

